design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,480 @@
+
+Once the migration completes, the controller will update the migrated volumes first in the VMI spec. This will trigger the update of the volumes in the VM spec if it exists.
+
+DataVolumes are taken into account in the current proposal. DataVolumes are used to import the image during VM provisioning, but they may essentially be thought of as PVCs for the rest of the VM life cycle. According to this proposal, KubeVirt will replace the source DataVolumes with the destination PVC when a user wishes to migrate data to a PVC.
+
+A schematic representation of the migration flow is shown below.
+![](volume-migration.png)
+
+### API examples
+
+Example of VolumeMigration flow
+
+Create a VM with source storage and the destination PVC:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: dest-pvc
+spec:
+  accessModes:
+    - ReadWriteOnce
+  volumeMode: Filesystem
+  resources:
+    requests:
+      storage: 1Gi
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  annotations:
+    kubevirt.io/libvirt-log-filters: "1:qemu.qemu_monitor 1:*"
+  labels:
+    kubevirt.io/vm: vm-alpine-datavolume
+  name: vm-alpine-datavolume
+spec:
+  dataVolumeTemplates:
+  - metadata:
+      creationTimestamp: null
+      name: source-pvc
+    spec:
+      pvc:
+        accessModes:
+        - ReadWriteOnce
+        resources:
+          requests:
+            storage: 1Gi
+        storageClassName: local
+      source:
+        registry:
+          url: docker://registry:5000/kubevirt/alpine-container-disk-demo:devel
+  running: true
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-alpine-datavolume
+    spec:
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: virtio
+            name: datavolumedisk1
+          interfaces:
+          - masquerade: {}
+            name: default
+        resources:
+          requests:
+            memory: 128Mi
+      networks:
+      - name: default
+        pod: {}
+      terminationGracePeriodSeconds: 0
+      volumes:
+      - dataVolume:
+          name: source-pvc
+        name: datavolumedisk1
+
+```
+
+Trigger the migration with:
+```yaml
+apiVersion: storage.kubevirt.io/v1alpha1
+kind: VolumeMigration
+metadata:
+  name: vol-mig
+spec:
+  migrationVolumes:
+  - sourcePvc: source-pvc
+    destinationPvc: dest-pvc
+    reclaimPolicySourcePvc: Delete
+```
+
+Once the migration succeed:
+
+Changes on the VMI:
+```yaml
+
+apiVersion: v1
+items:
+- apiVersion: kubevirt.io/v1
+  kind: VirtualMachineInstance
+[...]
+    volumes:
+    - name: datavolumedisk1
+      persistentVolumeClaim:
+        claimName: dest-pvc
+ status:
+    migrationState:
+      completed: true
+      endTimestamp: "2023-08-17T08:14:45Z"
+ [...]
+      migrationVolumes:
+      - destinationPvc: dest-pvc
+        sourcePvc: source-pvc
+        reclaimPolicySourcePvc: Delete
```

Comment:
Not sure about all this VolumeMigration info mirrored in the VMI.  Is it necessary?

Reply:
We could skip the reclaim policy, but the source and destination are necessary because virt-launcher uses this information on the VMI status to manipulate the migration XML and to understand which volumes need to be migrated.

Reply:
Not only virt-launcher uses this information, but also the VM and VMI controllers for update the VM/VMI spec. See the PR for reference: https://github.com/kubevirt/kubevirt/pull/10536

Reply:
Okay sounds good, was mostly wondering